torch/csrc/distributed/autograd/utils.cpp

```diff
@@ -115,14 +115,24 @@ std::shared_ptr<FutureMessage> sendMessageWithAutograd(
     RpcAgent& agent,
     const WorkerInfo& dst,
     torch::distributed::rpc::Message&& wrappedRpcMsg,
-    bool forceGradRecording) {
+    bool forceGradRecording,
+    std::shared_ptr<torch::autograd::profiler::RecordFunction> rf) {
   auto msg = getMessageWithAutograd(
       dst.id_,
       std::move(wrappedRpcMsg),
       MessageType::FORWARD_AUTOGRAD_REQ,
       forceGradRecording);
 
-  return agent.send(dst, std::move(msg));
+  auto fut = agent.send(dst, std::move(msg));
+  if (rf != nullptr) {
+    // save the local threadId so that end() callbacks can be correctly invoked
+    // from a different thread.
+    rf->setThreadId();
+    // attach the recordFunction object to the future, so callbacks can be
+    // invoked when the future finishes.
+    fut->attachRecordFunction(std::move(rf));
```

Comment:
What is the design reasoning for adding the `recordFunction` to `Future`? Is it because `Future` is what we get (about to get) from autograd and send? I am wondering whether we should introduce profiler concepts into the future or should we compose the future and the rf into a new type?

No need to make changes to the code yet, I am just trying to picking up design thinking here.

Reply:
For profiling, we'd like to run certain callbacks (such as the one that records how long the function/RPC took) when the future is completed. For example, if worker A sends an `rpc_async` to B, then in order for the correct time taken to appear we'd need to run the profiler callbacks when the future associated with the rpc is marked as done. So to do this we pass in the `RecordFunction` to the future.

I'd prefer not introducing another type that composes the RF + future, since that would require us to use this type everywhere we mark a future as completed, since it is possible that profiling might be enabled. Basically that type would be an RF + an optional RecordFunction, and the implementation details would leak into the RPC agents, which is not ideal (although, it is already not ideal that the profiler details leak into the future)

Reply:
> For profiling, we'd like to run certain callbacks (such as the one that records how long the function/RPC took) when the future is completed. 

If all we need is triggering callbacks when the future is done, can we implement the callback in the way that it binds a RF? So that Future does not need to have an RF field?

Reply:
I implemented the suggestion of doing it via a callback on the future, so now there are no changes to the generic Future.